File: day07.py
#!/usr/bin/env pipenv run python
"""Solutions to day 7 of Advent of Code"""

import re
import itertools
import collections

from get_input import get_input, line_parser

Record = collections.namedtuple('Record', 'before after')

def helper(records):
    result = {}
    for record in records:
        if record.before not in result:
            result[record.before] = set()
        if record.after not in result:
            result[record.after] = set()
        result[record.before].add(record.after)
    return result

def part1(records):
    """Solution to part 1"""
    solution = ""
    todo = helper(records)
    while todo:
        ready = list(k for k in todo.keys() if all(k not in v for v in todo.values()))
        ready.sort()
        solution += ready[0]
        del todo[ready[0]]
    return solution

def part2(records, n_workers=5, base=60):
    """Solution to part 2"""
    todo = helper(records)
    workers = {}
    time = 0
    while len(workers) != 0 or len(todo) != 0:
        t_diff = min((t for t, _ in workers.values()), default=0) + 1
        time += t_diff
        for jid in tuple(workers):
            workers[jid][0] -= t_diff 
            if workers[jid][0] < 0:
                del workers[jid]
        ready = list(
                k for k in todo.keys() if 
                all(k not in v for _, v in workers.values()) and
                all(k not in v for v in todo.values()))
        ready.sort()
        while len(workers) < n_workers and ready:
            item = ready.pop(0)
            workers[item] = [ord(item) - ord('A') + base, todo[item]]
            del todo[item]
    return time - 1

def parse(line):
    m = re.match(r"Step (.) must be finished before step (.) can begin.", line)
    return Record(*m.groups())

if __name__ == '__main__':
    LINES = line_parser(get_input(day=7, year=2018), parse=parse)
    print("Part 1: {}".format(part1(LINES)))
    print("Part 2: {}".format(part2(LINES)))

File: day07.rs
extern crate common;

use common::get_input;
use regex::Regex;

use std::collections::{HashMap, HashSet};

type Input = HashMap<String, HashSet<String>>;

fn part1(input: &Input) -> String {
    let mut todo = input.clone();
    let mut result: Vec<String> = Vec::new();
    while !todo.is_empty() {
        let mut to_remove: Vec<String> = Vec::new();
        for (key, values) in todo.clone() {
            if values.iter().all(|v| !todo.contains_key(v)) {
                to_remove.push(key);
            }
        }
        to_remove.sort();
        todo.remove(&to_remove[0]);
        result.push(to_remove[0].clone());
    }
    return result.join("");
}

fn part2(input: &Input) -> u32 {
    let mut todo = input.clone();
    let mut time = 0;
    let mut workers: Vec<(u32, String)> = Vec::new();
    while !todo.is_empty() {
        let t_diff = workers.iter().map(|v| v.0).min().unwrap_or(0) + 1;
        time += t_diff;
        workers = workers.iter().filter_map( |w| {
            if w.0 <= t_diff {
                todo.remove(&w.1);
                None
            } else {
                Some((w.0 - t_diff, w.1.clone()))
            }
        }).collect();
        let mut ready: Vec<String> = Vec::new();
        for key in todo.keys() {
            if todo.values().any(|vals| vals.contains(key)) {
                continue;
            } else if workers.iter()
                .any(|v| &v.1 == key) {
                continue
            }
            ready.push(key.clone());
        }
        ready.sort();
        while workers.len() < 5 && !ready.is_empty() {
            let item = ready.remove(0);
            let first = item.chars().next().unwrap() as u8;
            workers.push(((first - 'A' as u8 + 60) as u32, item.clone()));
        }
    }
    time
}

fn parse(lines: &String) -> Input {
    let mut result = HashMap::new();
    let format = Regex::new("Step (.) must be finished before step (.) can begin.")
        .unwrap();
    for line in lines.trim().split('\n') {
        let groups = format.captures(line).unwrap(); 
        result.entry(groups[1].to_string()).or_insert(HashSet::new());
        result.entry(groups[2].to_string())
            .or_insert(HashSet::new())
            .insert(groups[1].to_string());
    }
    return result
}

fn main() {
    let input = parse(&get_input(07, 2018));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
